### DISCUSSION TITLE
[FEAT] Visualizing domain architectures | network

### DISCUSSION THREAD

**Discussion by jananiravi** [JRaviLab/MolEvolvR#19] [Ideas]
### Implementing networks for Genomic Contexts (GCs) #23 
- [ ] Add network for genomic contexts as well, not just DA (split by arrow and ||)
- [ ] Include domains/DAs with checkboxes of those shown in the network so that they can be included/excluded by the user. Corresponding edges will appear/disappear too.
- [ ] Selection box for selecting domains/GCs to view (also applicable to table)
- [x] Include network plot and word cloud in the same tab
- [ ] add legend
- [ ] subset by lineage
- [ ] color node by lineage/...?
- [ ] pie chart lineage distribution overlaid
- [ ] network clustering
- [x] best igraph layout
- [x] in addition to `igraph` customizations, try other packages
  - [ ] https://github.com/grabear/awesome-rshiny
  - [ ] `ggraph` + `tidytext`
  - [ ] `ggnetwork`
  - [x] `networkD3`
  - [x] `visNetwork`
  - [ ] `geomnet`
  - [ ] `tidygraph`
  - [ ] [eg1, network analysis](https://www.jessesadler.com/post/network-analysis-with-r/) | [eg2, word relationships & ngrams](https://uc-r.github.io/word_relationships) | [eg3, ngrams](https://www.tidytextmining.com/ngrams.html)
- [ ] add more space for the network plot

**Comment by leenaaize** [comment]
Hello @jananiravi , I am an Outreachy applicant and I am interested in working on this

**Comment by jananiravi** [comment]
Sure, @leenaaize, please get started! share your plan of action once you go through the specifics.

**Reply by leenaaize** [reply]
Thank you @jananiravi, I'm currently working on my plan of action

**Comment by leenaaize** [comment]
Hello @jananiravi Here is my plan of action

Install necessary R packages:
igraph- To create network. 
shiny- To create shiny app.
ggraph - To visualize network and add customizations.
ggforce - To overlay pie charts on the nodes. 
dplyr - For data manipulation.
wordcloud2- To create word cloud.

Create Node List: Prepare the node list, containing unique GC and DA names, lineage information, and a unique identifier. Calculate lineage proportion for each node which will be used to create pie chart overlays.
 
Create Edge List: Prepare the edge list connecting GCs and DAs. The edge list will consist of from and to columns indicating the GC or DA relationships. 

Create igraph Object from node and edge lists.

Create Clusters:
Use clustering detection algorithms from the igraph package to identify clusters in the graph.
Extract cluster membership and add it to the node list, creating a new column for cluster IDs.

Design a Shiny App UI: Define the fluidPage() for the UI with sidebarPanel() and mainPanel(). The sidebar will include:
1. A checkbox group for users to select domains (GCs or DAs) to  include/exclude from the network plot.
2. A selection box  for users to choose whether to view Domains or Genomic Contexts in the network.
3. A checkbox group for selecting which lineages to filter the network by.
The main panel will display the network plot and word cloud side by side. Within the mainPanel function, placeholders will be defined for the plots that will be created in the server.

 Define the Shiny App Server logic:
1. Filter Nodes and Edges: Dynamically filter the nodes and edges based on user input (for domains, GCs, and lineages), ensuring the igraph object is updated accordingly using a reactive function.
2. Create Network Plot: Pass filtered igraph object into ggraph() and try various igraph layouts  to see which is best for the network. Use different shapes for nodes based on cluster IDs. Use different colors to map node color to lineage. Create legend for both color(lineage) and shape(clusters) ensuring these legends dynamically update when users filter the network. Use lineage proportion of each node to create piechart lineage distribution overlaid.
3. Create Word Cloud plot : Create a word cloud visualization using the wordcloud2 package, with words representing node labels (GCs/DAs).The word cloud will update based on the filtered nodes in the network plot.

Explore other packages for customizations such as visNetwork and networkD3 to add interactive features like zoom and drag.

 Adjust the size of the plot in Shiny using  the height and width  parameters in plotOutput().

Also just to clarify, when you specified that the network will be split by arrow and ||, does that mean there will be different types of relatonships, some directional  and some parallel?

**Comment by jananiravi** [comment]
This is a great start, @leenaaize!
Can you take a look at these scripts before getting started?
https://github.com/JRaviLab/MolEvolvR/blob/main/R/networks_domarch.R
https://github.com/JRaviLab/MolEvolvR/blob/main/R/networks_gencontext.R (currently unused)

Also, take a sneak peek into the MolEvolvR [web-app](//jravilab.org/molevolvr) domain architecture network tab to see how it's implemented in Shiny.

When going across the suggested packages above or other new ones in town, it's worth thinking about which features we would like, e.g., how interactive it is, SVG download options.

cc: @epbrenner @the-mayer

**Reply by leenaaize** [reply]
Alright I'm on it!

**Reply by leenaaize** [reply]
Hello @jananiravi please how do I get access to  the prot data used in the scripts above?
